design/dedicated-osd-pod.md

```diff
@@ -0,0 +1,58 @@
+# Run OSD in its own Pod
+
+## TL;DR
+
+A one-OSD-per-Pod placement should be implemented to improve reliability and resource efficiency for Ceph OSD daemon.
+
+## Background
+
+Currently in Rook 0.7, Rook Operator starts a ReplicaSet for Rook OSD agent on each storage node. The ReplicaSet has just one replica. Rook OSD agents scan and prepare devices, create OSD IDs and data directories, generate Ceph configuration. At last, OSD agents start all Ceph OSD daemons in foreground and track the OSD processes.
+
+As observed, all Ceph OSDs are running in the same Pod.
+
+## Limitations
+
+One Pod for all OSDs doesn't have the highest reliability nor efficiency. If the Pod is deleted, accidentally or during maintenence, all OSDs are down till the ReplicaSet restart. If nodes have asymmetric configurations, some have more devices than others, the Rook Operator doesn't request appropriate resources for the OSD Pods, since the resources are requested before the OSDs are discovered.
+
+A more comprehensive discussions can be found at [this issue](https://github.com/rook/rook/issues/1341).
+
+## Proposal   
+
+Based on offline discussion with [Travis](https://github.com/travisn), we propose the following OSD startup change.
+
+### Create new OSDs
+| Sequence |Rook Operator  | Rook OSD Agent  | Ceph OSD  | Note  |
+|---|---|---|---|---|
+| 1  | Create a Pod for OSD Agents for each storage node  |   |   | Consider a CronJob for peridically invocation to detect device change  |
+| 2  |   |  Discover and prepare OSDs  |   | |
```

Comment:
The preparation of the OSDs will need to be delayed until the operator decides that the device should be consumed (specified in the CRD). The provisioning step could then be done to prepare the device and create the osd id, likely this would happen as a deployment per device. Then finally the operator would start the deployment for the osd.

Reply:
Can you provide more details about how the discoverability of devices and preparation of OSDs will happen? I am wondering if we could have the agents leverage the K8S [LocalVolume](https://kubernetes.io/docs/concepts/storage/volumes/#local) to discover and present "available devices". LocalVolume is beta and enabled by default on 1.10. 
For example, if a directory/device is to be presented as a Rook device, it can be represented as a LocalVolume resource and the Operator can watch these LocalVolume objects to create the required resources like pods.

Doing this way, you could avoid using ConfigMaps and, instead, use LocalVolume to convey more information about the device using labels.